services/py-api/src/server/middleware/auth.py

```diff
@@ -0,0 +1,97 @@
+            if is_bypassed:
+                response = await call_next(request)
+                return response
+
+            # Provides way of authorizing the incoming request based on the environment
+            if os.environ["ENV"] != "PROD":
+                if request.headers.get("X-Auth-Token") == "DEV":
+                    response = await call_next(request)
+                    return response
+                else:
+                    return JSONResponse(content={"message": "Unauthorized"}, status_code=401)
+            else:
+                # TODO: Implement token authentication logic if we decide on an admin panel
+                # For now every effort to access authenticated routes in a PROD env will not be authorized!
+                return JSONResponse(content={"message": "Unauthorized"}, status_code=401)
+
+    @classmethod
+    def _check_bypassed_request(
+        cls, url_path: str, request_method: str
+    ) -> Tuple[str, Literal[True]] | Tuple[None, Literal[False]]:
+        # TODO: Strip the url from path params where there are such
+
+        # Check if the url_path matches any of the endpoints
+        endpoint, is_bypassed = cls._check_bypassed_endpoint(url_path)
+
+        # If we have a url_path match, check if the current 'request_method' matches any of the allowed url methods
+        if is_bypassed:
+            # If both the endpoint and methods are allowed, bypass request authorization
+            if request_method in cls._BYPASSED_REQUESTS[str(endpoint)]:
+                return url_path, True
+
+        return None, False
+
+    @classmethod
+    def _check_bypassed_endpoint(cls, url_path: str) -> Tuple[str, Literal[True]] | Tuple[None, Literal[False]]:
+
+        # Tokenize the url string using '/' as a delimiter
+        url_path_tokens = url_path.split("/")
+
+        for endpoint in cls._BYPASSED_REQUESTS.keys():
+            # Tokenize the endpoint string using '/' as a delimiter
+            endpoint_tokens = endpoint.split("/")
+
+            # Reset Sentinel Value/Flag to True
+            endpoints_match = True
+
+            # If the length of the endpoints doesnt match, they are not the same
+            # Skip whatever is below and continue to the next iteration
+            if len(url_path_tokens) != len(endpoint_tokens):
+                continue
+
+            for itr in range(len(endpoint_tokens)):
+                # Do not include the path parameters in the validation
+                # The regex expression checks for strings in the format '{...}'
+                if re.match(r"^\{.*\}$", endpoint_tokens[itr]):
+                    continue
+
+                # If there is one token mismatch the endpoints don't match
+                if endpoint_tokens[itr] != url_path_tokens[itr]:
+                    endpoints_match = False
+                    break
+
+            if endpoints_match:
+                return endpoint, True
```

Comment:
I think this logic is too complex. Wouldn't it be better to have something like this, if you want to have a way to allow path params:
```python
    _BYPASSED_REQUESTS: Final = {
        "/hackathon/teams/*": ["GET", "POST"]
        # We allow GET and POST for all types of path params, for example:
        # POST /hackathon/teams/id will pass
        # GET /hackathon/teams/isak/test will pass
    }
```
I can't understand why you split in two places and this whole logic with the tokens. Just compare the two strings, this will be much easier.

Here is my proposal:

```python
from os import environ
from typing import Callable, Any, Final

from fastapi.applications import FastAPI
from starlette.requests import Request
from starlette.responses import JSONResponse


class AuthMiddleware:
    """This middleware is responsible for handling authorization for incoming requests."""

    _BYPASSED_ENDPOINTS: Final = {
        "/ping": ["GET"],
        "/docs": ["GET"],
        "/openapi.json": ["GET"],
        "/hackathon/participants": ["POST"],
    }
    """
    This dictionary maps endpoint paths to allowed HTTP methods. It contains endpoints which do not require Bearer token
    to access (public endpoints).

    Paths ending with `/*` act as wildcards, matching any sub-path. For example:

    - `"/hackathon/teams/*": ["GET", "POST"]` matches paths like `/hackathon/teams/id` or `/hackathon/teams/isak/test` 
    with the specified methods (GET, POST).
    - `"/hackathon/participants": ["POST"]` matches only `/hackathon/participants` with the POST method.
    """

    def __init__(self, app: FastAPI) -> None:

        _base_url: str = app.root_path

        @app.middleware("http")
        async def verify_request(request: Request, call_next: Callable[[Any], Any]) -> JSONResponse:
            """
            Verifies the incoming request for authorization:
            - If the request is made to a bypassed endpoint, it is allowed to proceed.
            - If not, it checks for a Bearer token in the request headers for authorization.
            - If neither condition is met, it returns a 401 Unauthorized status code.

            Args:
                request: The incoming HTTP request.
                call_next: The next handler in the stack.
                    Passing the request to the next handler, either the next middleware in the stack or a route which
                    matches the URL pattern is handled internally by the `middleware` decorator

            For more info visit:
            https://fastapi.tiangolo.com/tutorial/middleware/
            https://www.starlette.io/middleware/
            https://github.com/encode/starlette/blob/8a99adfb5839b37efee33f2f49c1ba27a954fcbd/starlette/middleware/base.py#L108
            """
            if _is_bypassed(request):
                return await call_next(request)

            # FIXME: Currently the token is hardcoded as a secret for DEV and TEST environments, we will fix it before
            #  deploying to PROD
            if environ["ENV"] != "PROD":
                auth_header = request.headers.get("Authorization")
                if auth_header and auth_header.startswith("Bearer ") and auth_header[len("Bearer "):] == environ[
                    "SECRET_AUTH_TOKEN"]:
                    return await call_next(request)

                return JSONResponse(content={"message": "Unauthorized"}, status_code=401)

            # TODO: Implement Bearer token authorization logic if we decide on an admin panel.
            #  For now every effort to access protected routes in a PROD env will not be authorized!
            return JSONResponse(content={"message": "Unauthorized"}, status_code=401)

        def _is_bypassed(req: Request) -> bool:
            """Checks if the request towards the bypassed endpoints.

            Returns:
                True: if the request is made towards a bypassed endpoint
                False: otherwise
            """
            path = req.url.path.removeprefix(_base_url)
            method = req.method

            for allowed_path, allowed_methods in self._BYPASSED_ENDPOINTS.items():
                # Check for exact match
                if allowed_path == path and method in allowed_methods:
                    return True

                # Check for wildcard match
                if allowed_path.endswith("/*") and path.startswith(allowed_path[:-2]) and method in allowed_methods:
                    return True

            return False

```

Reply:
I was taking into account that the path parameters can be anywhere in the string that is why I took the approach of the tokens. I was accounting for /path/*/with/multiple/*/path/parameters/*
I agree with you that it should be simpler. Having multiple path parameter is a rare case. I don't know what I was thinking lol, I just got to carried away in making it as flexible as possible

Reply:
However there is a flow in the logic. every possible path ends with /* .How will you

> I was taking into account that the path parameters can be anywhere in the string that is why I took the approach of the tokens. I was accounting for "/path/*/with/multiple/*/path/parameters/*" I agree with you that it should be simpler. Having multiple path parameter is a rare case. I don't know what I was thinking lol, I just got to carried away in making it as flexible as possible

Reply:
Okayyyy. Why don't you look into auth middleware as a decorator?

```python
@permissions(TEAM.ADMIN)
def remove_team_participant(...)
```

Reply:
Here's what I found from a quick search

https://github.com/code-specialist/fastapi-auth-middleware

Reply:
@asynchroza Thanks for the suggestion look much more elegant!

Reply:
Why writing your own tooling when you can just someone else's stuff😃

Reply:
@asynchroza @Isak-Bego, So finally, we are using decorators, as they are more flexible?

Reply:
> Why bother writing your own tooling when you can just use someone else's stuff

As you know, the saying is "do not reinvent the wheel"

> Didn't you hate decorators?

No. They're a good tool as long as not abused. Yes, you should be mindful of introducing too much indirection.

Reply:
> why you suggested this approach?

I suggest using well-established and proven methods rather than creating unconventional regex patterns to match your paths. Additionally, defining scopes at the top of the controllers improves clarity and maintains good locality.

Reply:
Fixed in 081ff0183b99110eff78fabb6c982deb88ae3869